### DISCUSSION TITLE
TTree.arrays() first argument with paths ("/" character meant path in Uproot 3, division in Uproot 4)

### DISCUSSION THREAD

**Discussion by ast0815** [scikit-hep/uproot5#283] [Q&A]
Hello, I am currently testing uproot4 functions with the Histoprint command line tool. In uproot3, one could load branches without specifying the full path. In uproot4 that does not seem to be possible.

```python
>>> import uproot
>>> uproot.__version__
4.0.4
```

```
$ wget --quiet http://scikit-hep.org/uproot3/examples/Event.root
$ python -c "import uproot as up; up.open('Event.root')['T'].arrays(['event/fTracks.fYfirst'])"
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/home/koch/Projekte/histoprint/env/lib/python3.7/site-packages/uproot/behaviors/TBranch.py", line 1133, in arrays
    arrays,
  File "/home/koch/Projekte/histoprint/env/lib/python3.7/site-packages/uproot/behaviors/TBranch.py", line 3460, in _ranges_or_baskets_to_arrays
    uproot.source.futures.delayed_raise(*obj)
  File "/home/koch/Projekte/histoprint/env/lib/python3.7/site-packages/uproot/source/futures.py", line 46, in delayed_raise
    raise exception_value.with_traceback(traceback)
  File "/home/koch/Projekte/histoprint/env/lib/python3.7/site-packages/uproot/behaviors/TBranch.py", line 3414, in basket_to_array
    library,
  File "/home/koch/Projekte/histoprint/env/lib/python3.7/site-packages/uproot/interpretation/objects.py", line 141, in basket_array
    form = self.awkward_form(branch.file, index_format="i64")
  File "/home/koch/Projekte/histoprint/env/lib/python3.7/site-packages/uproot/interpretation/objects.py", line 122, in awkward_form
    self._branch.file, index_format, header, tobject_header, breadcrumbs
  File "/home/koch/Projekte/histoprint/env/lib/python3.7/site-packages/uproot/containers.py", line 645, in awkward_form
    self._values, file, index_format, header, tobject_header, breadcrumbs
  File "/home/koch/Projekte/histoprint/env/lib/python3.7/site-packages/uproot/_util.py", line 426, in awkward_form
    file, index_format, header, tobject_header, breadcrumbs
  File "/home/koch/Projekte/histoprint/env/lib/python3.7/site-packages/uproot/containers.py", line 369, in awkward_form
    raise uproot.interpretation.objects.CannotBeAwkward(self.message)
uproot.interpretation.objects.CannotBeAwkward: Double32_t in array (note: Event.root fClosestDistance has an example)
```

This used to work with uproot3. It also does work with uproot4 when explicitly providing the full path to the variable:

```
$ python -c "import uproot as up; up.open('Event.root')['T'].arrays(['event/fTracks/fTracks.fYfirst'])"
```

Not the additional `/fTracks`.

Easy to fix by providing the full paths, but I am not sure whether this is expected behaviour.

**Comment by ast0815** [comment]
I spoke too soon. When trying to address specific indices, it fails again:

```
$ python -c "import uproot as up; up.open('Event.root')['T'].arrays(['event/fTracks/fTracks.fYfirst[...:0]'])"
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/home/koch/Projekte/histoprint/env/lib/python3.7/site-packages/uproot/behaviors/TBranch.py", line 1133, in arrays
    arrays,
  File "/home/koch/Projekte/histoprint/env/lib/python3.7/site-packages/uproot/behaviors/TBranch.py", line 3460, in _ranges_or_baskets_to_arrays
    uproot.source.futures.delayed_raise(*obj)
  File "/home/koch/Projekte/histoprint/env/lib/python3.7/site-packages/uproot/source/futures.py", line 46, in delayed_raise
    raise exception_value.with_traceback(traceback)
  File "/home/koch/Projekte/histoprint/env/lib/python3.7/site-packages/uproot/behaviors/TBranch.py", line 3414, in basket_to_array
    library,
  File "/home/koch/Projekte/histoprint/env/lib/python3.7/site-packages/uproot/interpretation/objects.py", line 141, in basket_array
    form = self.awkward_form(branch.file, index_format="i64")
  File "/home/koch/Projekte/histoprint/env/lib/python3.7/site-packages/uproot/interpretation/objects.py", line 122, in awkward_form
    self._branch.file, index_format, header, tobject_header, breadcrumbs
  File "/home/koch/Projekte/histoprint/env/lib/python3.7/site-packages/uproot/containers.py", line 645, in awkward_form
    self._values, file, index_format, header, tobject_header, breadcrumbs
  File "/home/koch/Projekte/histoprint/env/lib/python3.7/site-packages/uproot/_util.py", line 426, in awkward_form
    file, index_format, header, tobject_header, breadcrumbs
  File "/home/koch/Projekte/histoprint/env/lib/python3.7/site-packages/uproot/containers.py", line 369, in awkward_form
    raise uproot.interpretation.objects.CannotBeAwkward(self.message)
uproot.interpretation.objects.CannotBeAwkward: Double32_t in array (note: Event.root fClosestDistance has an example)
```

Slicing does work for another file I have though: [test.root.zip](https://github.com/scikit-hep/uproot4/files/6022713/test.root.zip)

```
$ python -c "import uproot as up; up.open('test.root')['truth'].arrays(['truelepton_dir[...,0]'])"
```

**Comment by jpivarski** [comment]
Full paths or relative paths are not the issue: what it's trying to do here is load the whole `event` array as a single object to _divide it_ by `fTracks`, and _divide that_ by `fTracks.fYfirst[...:0]`. The first argument of [uproot.TTree.arrays](https://uproot.readthedocs.io/en/latest/uproot.behaviors.TTree.TTree.html#arrays) is "`expressions`", quantities that are computed by Python. (You are using the fact that this is a computed expression by slicing `fTracks.fYfirst` with `..., :0`, which couldn't be done in Uproot 3.)

The nested path names (with `/` not meaning "divide") and pattern matching (with `*` not meaning "multiply") are in the `filter_name` parameter. What you meant to say in your first example is:

```python
>>> uproot.open("/mnt/storage/data/misc/Event.root")["T"].arrays(
...     ["fTracks.fYfirst[..., :0]"],
...     filter_name="event/fTracks/fTracks.fYfirst")
<Array [{'fTracks.fYfirst[..., :0]': [, ... ] type='1000 * {"fTracks.fYfirst[......'>
```

or you could hide that unwieldy field name as an alias (which work just like ROOT's built-in fAliases):

```python
>>> uproot.open("/mnt/storage/data/misc/Event.root")["T"].arrays(
...     ["first"],
...     filter_name="event/fTracks/fTracks.fYfirst",
...     aliases={"first": "fTracks.fYfirst[..., :0]"})
<Array [{first: []}, ... {first: []}] type='1000 * {"first": var * float32}'>
```

Also, if you're after just one branch, you can navigate to that branch explicitly and avoid all this searching and renaming.

```python
>>> uproot.open("/mnt/storage/data/misc/Event.root")["T/event/fTracks/fTracks.fYfirst"].array()[..., :0]
<Array [[], [], [], [], ... [], [], [], []] type='1000 * var * float32'>
```

Now I'm wondering if you really meant to slice it with "`..., :0`", since this would give you all empty lists, but the slicing is independent of the distinction between `expressions` (computable expressions, possibly aliased), `filter_name` (picking out the branches you want), and paths (navigating through TDirectories and TBranches).

I don't see a bug here, and the above would be good for everybody to know, so I'm going to make this a Discussion. If I'm wrong and there's still a bug, open a new Issue. (Unfortunately, we can't turn Discussions back into Issues, but a second iteration on your problem would likely be more specific, anyway, justifying a new Issue.)

**Reply by ast0815** [reply]
Right, there was a typo there. It should have been `[:,0]`, i.e. selecting only the first element for all events, like in the second example.

Let me explain my use case, and maybe you have a suggestion how to bes implement this:

The `histoprint` CLI program can read in ROOT files and plot histograms from TTree branches. It used to get the data of the branches one-by-one, like you suggested. I would like to extend that functionality by allowing users to provide a "cut" so that the data can be filtered before plotting.

Now when reading the uprrot3->4 cheat sheet, I saw that the `arrays` method accepts a `cut` argument, while the `array` (singular) method does not. I assumed that the `arrays` method would thus be more efficient by doing the cutting while reading the TTree rather than me loading branches and then calculating a boolean index to do the cutting. If you say that the evaluation of the first two arguments happens on Python objects, is there no performance gain from using `arrays` rather than `array`?

What would you suggest is the best way to load multiple branches (potentially with indices, e.g. to plot the 2nd component of a direction vector) with some sort of filtering based on those (or other) branches?

So far I have kept the "provide full path" approach, because some ROOT files might have sub-branches with identical names.

**Reply by jpivarski** [reply]
There is no performance gain from using strings (e.g. non-trivial `expressions` or `cut`), rather than working with Python objects. In fact, that's how the string-evaluation is implemented.

The reason it exists at all is to implement `aliases`. Those `aliases` are usually expressed in TTreeFormula syntax, which is not the same as Awkward Arrays in Python, so the job isn't entirely done: we'll need a parser for TTreeFormula so that this can be passed as the `language` argument (probably by default, after a switchover period). But when that's complete, there will still be this issue about `/` meaning division and `*` meaning multiplication that we need to get used to now.

There isn't a way to improve performance beyond reading the arrays used in the cut and the arrays they are meant to cut, then performing that calculation array-wise. TBaskets can't be partially read because they're compressed as a single block, in order to read any byte from a TBasket, you have to read the whole TBasket. There is one caveat: you could do a little better than "read TBaskets into whole arrays, apply cut" by "read individual TBaskets, apply cut to each TBasket-array, then combine TBasket-arrays into whole arrays," which saves a memory copy, but most calculations you might want to do in order to implement the cut are more expensive than a memory-copy.

You can still provide the whole path, but it has to be in `filter_name` (which can be a list, if you're loading several), not `expressions`.
